From eb4f73adb865cd668577b021546f9323862cd4e0 Mon Sep 17 00:00:00 2001
From: Alexandre Gonzalo <alexandre.gonzalo@trusted-logic.com>
Date: Mon, 28 Mar 2011 10:23:54 -0500
Subject: OMAP4: SMC: CPU1 off while configuring EMIF FW with CPU0

CPU1 must be not accesing to EMIF while CPU0 configures
EMIF firewall.

Specific EMIF FW setting is needed while SMC PA is loaded.

CPU1 must not access EMIF while CPU0 configures EMIF firewall.
Otherwise, if CPU1 accesses to EMIF while busy bit is not
released (due to CPU0 configuring EMIF FW), then system hang
might occur (HW Bug:FW_NTTP-FAMILYBUG00006)

SMC Bug ID:6082

Change-Id: I48a700cb01ba68338daf2d44a7ed20b45ee434f0
Signed-off-by: Alexandre Gonzalo <alexandre.gonzalo@trusted-logic.com>
Signed-off-by: Rigaud Jean-Luc <jl-rigaud@ti.com>
Signed-off-by: Praneeth Bajjuri <praneeth@ti.com>
---
 security/smc/omap4/scxlnx_comm_mshield.c |   12 +++++++++++-
 1 files changed, 11 insertions(+), 1 deletions(-)

Index: kernel/security/smc/omap4/scxlnx_comm_mshield.c
===================================================================
--- kernel.orig/security/smc/omap4/scxlnx_comm_mshield.c	2011-04-07 00:01:55.000000000 +0900
+++ kernel/security/smc/omap4/scxlnx_comm_mshield.c	2011-04-07 02:27:07.000000000 +0900
@@ -20,7 +20,7 @@
 #include <asm/div64.h>
 #include <asm/system.h>
 #include <asm/cputype.h>
-#include <asm/uaccess.h>
+#include <linux/uaccess.h>
 #include <linux/io.h>
 #include <linux/interrupt.h>
 #include <linux/page-flags.h>
@@ -29,6 +29,7 @@
 #include <linux/version.h>
 #include <linux/jiffies.h>
 #include <linux/dma-mapping.h>
+#include <linux/cpu.h>
 
 #include <asm/cacheflush.h>
 
@@ -723,6 +724,9 @@
 	 */
 	SCXLNXCommSetCurrentTime(pComm);
 
+	/* Workaround for issue #6082 */
+	disable_nonboot_cpus();
+
 	/*
 	 * Start the SMC PA
 	 */
@@ -787,6 +791,8 @@
 	goto exit;
 
 error2:
+	/* Workaround for issue #6082 */
+	enable_nonboot_cpus();
 	spin_lock(&(pComm->lock));
 	pL1SharedBuffer = pComm->pBuffer;
 	pInitSharedBuffer = pComm->pInitSharedBuffer;
